`parent` keyword is getting scoped

## Bug report

| Question    | Answer
| ------------| ---------------
| PHP-Box version | 3.11.1
| PHP version | 7.2, 7.3, 7.4 (8.0 seems to work)
| Platform with version | Ubuntu
| Github Repo | [captainhookphp/captainhook](https://github.com/captainhookphp/captainhook)


During the scoping process of `Symfony\String` `5.2` classes that use `parent` as type hint the `parent` keyword  is getting  scoped to `PREFIX\parent`

```
PHP Fatal error:  Cannot use 'CaptainHook\Ext\parent' as class name as it is reserved in phar:///home/runner/work/captainhook/captainhook/build/phar/captainhook.phar/vendor/symfony/string/UnicodeString.php on line 166
16
Error: Process completed with exit code 255.
```
You can look at the full build result [here](https://github.com/captainhookphp/captainhook/runs/2231699832?check_suite_focus=true)

Original Symfony code
```php
public function normalize(int $form = self::NFC): parent
{
        $str = clone $this;
    ...
}
```

Scoped version
```php
public function normalize(int $form = self::NFC) : \CaptainHook\Ext\parent
{
    $str = clone $this;
    ...
}
```

The `parent` keyword should not be scoped.

Thanks :)

--- a/specs/special-keywords/self-parent-return-type.php
+++ b/specs/special-keywords/self-parent-return-type.php
@@ -1,0 +1,175 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the humbug/php-scoper package.
+ *
+ * Copyright (c) 2017 Théo FIDRY <theo.fidry@gmail.com>,
+ *                    Pádraic Brady <padraic.brady@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+return [
+    'meta' => [
+        'title' => 'Self and parent keywords as return types',
+        // Default values. If not specified will be the one used
+        'prefix' => 'Humbug',
+        'whitelist' => [],
+        'whitelist-global-constants' => true,
+        'whitelist-global-classes' => false,
+        'whitelist-global-functions' => true,
+        'registered-classes' => [],
+        'registered-functions' => [],
+    ],
+
+    'Usage for classes in the global scope' => <<<'PHP'
+<?php
+
+class A {
+    protected $name;
+    public function __construct(string $name)
+    {
+        $this->name = $name;
+    }
+    public function normalize() : self
+    {
+        $instance = clone $this;
+        $instance->name = \strtolower($this->name);
+        return $instance;
+    }
+    public function getName() : string
+    {
+        return $this->name;
+    }
+}
+
+class B extends A {
+    
+    public function normalize() : parent
+    {
+        $instance = clone $this;
+        $instance->name = \strtoupper($this->name);
+        return $instance;
+    }
+}
+
+echo (new B('yo'))->normalize()->getName().PHP_EOL;
+
+----
+<?php
+
+namespace Humbug;
+
+class A
+{
+    protected $name;
+    public function __construct(string $name)
+    {
+        $this->name = $name;
+    }
+    public function normalize() : self
+    {
+        $instance = clone $this;
+        $instance->name = \strtolower($this->name);
+        return $instance;
+    }
+    public function getName() : string
+    {
+        return $this->name;
+    }
+}
+class B extends \Humbug\A
+{
+    public function normalize() : parent
+    {
+        $instance = clone $this;
+        $instance->name = \strtoupper($this->name);
+        return $instance;
+    }
+}
+echo (new \Humbug\B('yo'))->normalize()->getName() . \PHP_EOL;
+
+PHP
+    ,
+
+    'Usage for classes in a namespaced' => <<<'PHP'
+<?php
+
+namespace Foo {
+    class A {
+        protected $name;
+        public function __construct(string $name)
+        {
+            $this->name = $name;
+        }
+        public function normalize() : self
+        {
+            $instance = clone $this;
+            $instance->name = \strtolower($this->name);
+            return $instance;
+        }
+        public function getName() : string
+        {
+            return $this->name;
+        }
+    }
+        
+    class B extends A {
+        public function normalize() : parent
+        {
+            $instance = clone $this;
+            $instance->name = strtoupper($this->name);
+            return $instance;
+        }
+    }
+}
+
+namespace {
+    use Foo\B;
+
+    echo (new B('yo'))->normalize()->getName().PHP_EOL;
+}
+
+----
+<?php
+
+namespace Humbug\Foo;
+
+class A
+{
+    protected $name;
+    public function __construct(string $name)
+    {
+        $this->name = $name;
+    }
+    public function normalize() : self
+    {
+        $instance = clone $this;
+        $instance->name = \strtolower($this->name);
+        return $instance;
+    }
+    public function getName() : string
+    {
+        return $this->name;
+    }
+}
+class B extends \Humbug\Foo\A
+{
+    public function normalize() : parent
+    {
+        $instance = clone $this;
+        $instance->name = \strtoupper($this->name);
+        return $instance;
+    }
+}
+namespace Humbug;
+
+use Humbug\Foo\B;
+echo (new \Humbug\Foo\B('yo'))->normalize()->getName() . \PHP_EOL;
+
+PHP
+    ,
+];

--- a/src/PhpParser/NodeVisitor/NameStmtPrefixer.php
+++ b/src/PhpParser/NodeVisitor/NameStmtPrefixer.php
@@ -194,7 +194,7 @@ final class NameStmtPrefixer extends NodeVisitorAbstract
             }
         }
 
-        if ('self' === (string) $resolvedName && $parentNode instanceof ClassMethod) {
+        if ($parentNode instanceof ClassMethod && $resolvedName->isSpecialClassName()) {
             return $name;
         }
 

